File: src/core/workshop.ts
import { z } from 'zod';
import type { WorkshopId } from './acf.ts';
import { AxeError } from './errors.ts';

const URL = 'https://api.steampowered.com/ISteamRemoteStorage/GetPublishedFileDetails/v1/';
const TIMEOUT_MS = 15_000;

const ItemRawSchema = z
  .object({
    publishedfileid: z.string(),
    title: z.string(),
    time_updated: z.number().int(),
    visibility: z.number().int(),
    result: z.number().int(),
  })
  .passthrough();

const ResponseSchema = z.object({
  response: z.object({
    publishedfiledetails: z.array(ItemRawSchema),
  }),
});

export type WorkshopItem = {
  published_file_id: WorkshopId;
  title: string;
  time_updated: number;
  visibility: number;
  result: number; // 1 = published; non-1 = errored / deleted
};

/** Function shape compatible with global `fetch`; loose on purpose so tests can supply minimal fakes. */
export type FetchLike = (input: string, init: RequestInit) => Promise<Response>;

export type GetDetailsOptions = {
  signal?: AbortSignal;
  fetchImpl?: FetchLike; // injection point for tests; not part of public contract
};

/**
 * Query the Workshop API for the given list of published file IDs.
 *
 * No API key required (public endpoint). 15s timeout. Returns typed items
 * with bigint IDs; throws AxeError(workshop_api) on HTTP / JSON / shape
 * failures.
 */
export async function getPublishedFileDetails(
  ids: readonly WorkshopId[],
  options: GetDetailsOptions = {},
): Promise<WorkshopItem[]> {
  if (ids.length === 0) return [];

  const body = new URLSearchParams();
  body.set('itemcount', String(ids.length));
  for (let i = 0; i < ids.length; i++) {
    body.set(`publishedfileids[${i}]`, (ids[i] as bigint).toString());
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);
  if (options.signal) {
    options.signal.addEventListener('abort', () => controller.abort(), { once: true });
  }

  const fetcher = options.fetchImpl ?? fetch;

  let response: Response;
  try {
    response = await fetcher(URL, {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body,
      signal: controller.signal,
    });
  } catch (e) {
    throw new AxeError('workshop_api', `HTTP request: ${errorMessage(e)}`);
  } finally {
    clearTimeout(timeout);
  }

  if (!response.ok) {
    throw new AxeError('workshop_api', `HTTP ${response.status} ${response.statusText}`);
  }

  let raw: unknown;
  try {
    raw = await response.json();
  } catch (e) {
    throw new AxeError('workshop_api', `parsing JSON: ${errorMessage(e)}`);
  }

  const parsed = ResponseSchema.safeParse(raw);
  if (!parsed.success) {
    throw new AxeError('workshop_api', `unexpected response shape: ${parsed.error.message}`);
  }

  return parsed.data.response.publishedfiledetails.map((item) => ({
    published_file_id: BigInt(item.publishedfileid),
    title: item.title,
    time_updated: item.time_updated,
    visibility: item.visibility,
    result: item.result,
  }));
}

function errorMessage(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

File: src/axe/workshop.py
"""Steam Workshop API client: `GetPublishedFileDetails` via httpx, no API key."""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass

import httpx
from pydantic import BaseModel, ValidationError

from axe.errors import AxeError

_URL = "https://api.steampowered.com/ISteamRemoteStorage/GetPublishedFileDetails/v1/"
_TIMEOUT_SECONDS = 15.0


class _RawItem(BaseModel):
    model_config = {"extra": "allow"}
    publishedfileid: str
    title: str
    time_updated: int
    visibility: int
    result: int


class _RawResponseInner(BaseModel):
    publishedfiledetails: list[_RawItem]


class _RawResponse(BaseModel):
    response: _RawResponseInner


@dataclass(frozen=True)
class WorkshopItem:
    published_file_id: int
    title: str
    time_updated: int
    visibility: int
    result: int  # 1 = published; non-1 = errored / deleted


FetchLike = Callable[[str, dict[str, str]], httpx.Response]


def _default_fetch(url: str, form: dict[str, str]) -> httpx.Response:
    return httpx.post(
        url,
        data=form,
        timeout=_TIMEOUT_SECONDS,
        headers={"content-type": "application/x-www-form-urlencoded"},
    )


def get_published_file_details(
    ids: Sequence[int],
    *,
    fetch: FetchLike | None = None,
) -> list[WorkshopItem]:
    """Query Workshop API for the given file IDs. Raises AxeError(workshop_api)."""
    if not ids:
        return []
    form: dict[str, str] = {"itemcount": str(len(ids))}
    for i, file_id in enumerate(ids):
        form[f"publishedfileids[{i}]"] = str(file_id)

    impl = fetch or _default_fetch
    try:
        response = impl(_URL, form)
    except httpx.HTTPError as e:
        raise AxeError("workshop_api", f"HTTP request: {e}") from e

    if response.status_code != 200:
        raise AxeError(
            "workshop_api", f"HTTP {response.status_code} {response.reason_phrase}"
        )

    try:
        raw = response.json()
    except ValueError as e:
        raise AxeError("workshop_api", f"parsing JSON: {e}") from e

    try:
        parsed = _RawResponse.model_validate(raw)
    except ValidationError as e:
        raise AxeError("workshop_api", f"unexpected response shape: {e}") from e

    return [
        WorkshopItem(
            published_file_id=int(item.publishedfileid),
            title=item.title,
            time_updated=item.time_updated,
            visibility=item.visibility,
            result=item.result,
        )
        for item in parsed.response.publishedfiledetails
    ]
